Public/Sdk/Public/Transformers/Transformer.Execute.dsc

```diff
@@ -161,6 +161,14 @@ namespace Transformer {
          * for a pip. The pip will still be stored to the cache for sake of supporting distribution.
          */
         disableCacheLookup?: boolean;
+
+        /**
+         * The weighted value of this pip when limiting concurrency of process pips.
+         * The default weight for a process pip is 1, and the max valid weight is the max concurrent processes allowed in the build.

```

Comment:
>a process pip is 1 [](start = 34, length = 18)

What's the meaning of pip with weight 0?

Reply:
- "weight" defaults to 1, converts values < 1 to 1, and considers values >= available process slots to mean the process should run alone

There are some tests to double check this behavior also.

---
In reply to: [271093105](https://github.com/Microsoft/BuildXL/pull/67#discussion_r271093105) [](ancestors = 271093105)

Reply:
I was thinking that the default weight should be 0, and the allowed range would be [0, 1). Then the weight can be computed something like 1/(1-v). I also don't think that we need to mention that the max valid weight is the max concurrent process because one can just change that max. The weight should be relative so to say.

---
In reply to: [271093105](https://github.com/Microsoft/BuildXL/pull/67#discussion_r271093105) [](ancestors = 271093105)

Reply:
When thinking about weight I'm a bit accustomed with [0, 1). Or something more like what Danny's mentioned above. Perhaps @almilli can give some insight.

---
In reply to: [271096384](https://github.com/Microsoft/BuildXL/pull/67#discussion_r271096384) [](ancestors = 271096384,271093105)

Reply:
How does one reason about the relative cost of a pip with a weight of 0.2 compared to a pip with a weight of 0?

The integer multiplier here is closely aligned with CloudBuild/QBuild/QTest weight concepts according to:
[https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2F1eswiki.com%2Fwiki%2FCBFAQ%23I_am_getting_ExceededMaxBuildTime_exception_while_running_buddy_build.2C_What_should_I_do_to_resolve_this_issue.3F&data=02%7C01%7CAnnie.Fu%40microsoft.com%7Cccbc5e5d4cb348c384ed08d6b22f7891%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636892314363790059&sdata=lTJrSIMbQ%2BSkva%2B7U%2Ba8RnDnxhRpi5zZB96LF8Dv0AU%3D&reserved=0](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2F1eswiki.com%2Fwiki%2FCBFAQ%23I_am_getting_ExceededMaxBuildTime_exception_while_running_buddy_build.2C_What_should_I_do_to_resolve_this_issue.3F&data=02%7C01%7CAnnie.Fu%40microsoft.com%7Cccbc5e5d4cb348c384ed08d6b22f7891%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636892314363790059&sdata=lTJrSIMbQ%2BSkva%2B7U%2Ba8RnDnxhRpi5zZB96LF8Dv0AU%3D&reserved=0) 


---
In reply to: [271097510](https://github.com/Microsoft/BuildXL/pull/67#discussion_r271097510) [](ancestors = 271097510,271096384,271093105)

Reply:
DScript doesn't support floating-point numbers, so, practically speaking, we should stick with integers here.

Reply:
After offline discussion, we are going to keep the semantics in-line with quickbuild, qtest, and cloudbuild and have it be an integer that defaults to 1.

---
In reply to: [271099495](https://github.com/Microsoft/BuildXL/pull/67#discussion_r271099495) [](ancestors = 271099495,271097510,271096384,271093105)

Reply:
I don't think we differentiate the "lightness" of lightweight pips right now, so I would say actually just using 0 as the value for lightweight pips might be the way to go.

We could eliminate the lightweight queue altogether and just schedule pips with a weight of 0, meaning those pips need 0 process slots and will always run unless restricted by some additional semaphore.

That being said, I am going to defer handling lightweight pips to outside of this PR.

---
In reply to: [271103676](https://github.com/Microsoft/BuildXL/pull/67#discussion_r271103676) [](ancestors = 271103676)